invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java

```diff
@@ -199,75 +201,78 @@ protected UsageInArrearAggregate getToBeBilledUsageDetails(final LocalDate start
         }
     }
 
-    List<UsageConsumableInArrearTierUnitAggregate> computeToBeBilledConsumableInArrearWith_ALL_TIERS(final LocalDate startDate, final LocalDate endDate, final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final Long units) throws CatalogApiException {
-
-        List<UsageConsumableInArrearTierUnitAggregate> toBeBilledDetails = Lists.newLinkedList();
-        long remainingUnits = units;
+    List<UsageConsumableInArrearTierUnitAggregate> computeToBeBilledConsumableInArrearWith_ALL_TIERS(final LocalDate startDate, final LocalDate endDate, final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final BigDecimal units) throws CatalogApiException {
+        final List<UsageConsumableInArrearTierUnitAggregate> toBeBilledDetails = Lists.newLinkedList();
+        BigDecimal remainingUnits = units;
         int tierNum = 0;
 
         final int lastPreviousUsageTier = previousUsage.size(); // we count tier from 1, 2, ...
         final boolean hasPreviousUsage = lastPreviousUsageTier > 0;
 
         for (final TieredBlock tieredBlock : tieredBlocks) {
-
             tierNum++;
-            final long blockTierSize = tieredBlock.getSize().longValue();
-            final long tmp = remainingUnits / blockTierSize + (remainingUnits % blockTierSize == 0 ? 0 : 1);
-            long nbUsedTierBlocks;
-            if (tieredBlock.getMax() != (double) -1 && tmp > tieredBlock.getMax() ) {
-                nbUsedTierBlocks = tieredBlock.getMax().longValue();
-                remainingUnits -= tieredBlock.getMax() * blockTierSize;
+            final BigDecimal blockTierSize = tieredBlock.getSize();
+            final BigDecimal blockTierMax = tieredBlock.getMax();
+            final long tmp = remainingUnits.longValue() / blockTierSize.longValue() + (remainingUnits.longValue() % blockTierSize.longValue() == 0 ? 0 : 1);
```

Comment:
I think we should also use the `BigDecimal#divideAndRemainder` instead of converting to long here - the way it's done a few lines below ?